stdlib/public/core/StringStorage.swift

```diff
@@ -66,6 +68,52 @@ private func _getCharacters<T:_AbstractStringStorage>(_ this:T,
     range: range)
 }
 
+@_effects(releasenone)
+private func _getCString<T:_AbstractStringStorage>(_ this: T,
+                             _ bufferPtr: UnsafeMutablePointer<UInt8>?,
+                             _ maxLength: Int,
+                             _ encoding: UInt)
+  -> Int8 {
+  let buffer:UnsafeMutableBufferPointer<UInt8>? = bufferPtr != nil ? UnsafeMutableBufferPointer(start: bufferPtr, count: maxLength) : nil
+  
+  switch (encoding, this.isASCII) {
+  case (_cocoaASCIIEncoding, true):
+    fallthrough
+  case (_cocoaUTF8Encoding, _):
+    if let buffer = buffer {
+      let utf8 = this.asString.utf8
+      if (maxLength < utf8.underestimatedCount) {
+        return 0
+      }
+      _ = utf8._copyContents(initializing: buffer)
+    }
+    return 1
+  default:
+    break
+  }
+  return  _cocoaGetCStringTrampoline(this,
+                                   bufferPtr,
+                                   maxLength,
+                                   encoding)
+}
```

Comment:
Seems like we can simplify the code a bit (following stdlib coding conventions):

```swift
@_effects(releasenone)
private func _getCString<T:_AbstractStringStorage>(
  _ this: T,
  _ outputPtr: UnsafeMutablePointer<UInt8>?,
  _ maxLength: Int,
  _ encoding: UInt
) -> Int8 {
  switch (encoding, this.isASCII) {
  case (_cocoaASCIIEncoding, true):
    fallthrough
  case (_cocoaUTF8Encoding, _):
    guard let outPtr = outputPtr else { return 1 }
    guard maxLength >= this.count else { return 0 }

    let buffer = UnsafeMutableBufferPointer(start: outPtr, count: maxLength)
    _ = this.asString.utf8._copyContents(initializing: buffer)
   }

   ...
 }
```

Reply:
Also, I'm probably misunderstanding the API contract when the pointer is nil. Should we always return if the buffer is nil or too small, and skip the fall back path?

Reply:
Looks good. And yes; actually you caught something else there, the buffer is not supposed to be optional for this one. If it's too small I believe we should just return, given that the NSString implementation passes NSFailOnPartialEncodingConversion to the underlying transcoding funnel point.

Reply:
(My rationale here is that we're already UTF8, so if we don't fit, then asking NSString to transcode us to UTF16 then back to UTF8 isn't going to make us fit)

Reply:
Why do we pass in the UTF8View? Given that we have a UnsafeBufferPointer<UInt8>, we should just copy directly rather than jumping through hoops. I think these helper methods can be simple `@inline(__always)` generic over the `AbstractStringStorage` class, and don't need any parameters you can't get from the `AbstractStringStorage`.

An alternative is to make a `inline(__always)` internal protocol and not do inheritance here. That protocol would require `count`, `start`, etc., and contain all the implementations in an extension. Then our two storage types trivially conform and no need for the abstract base class.